Imaging pipeline

paolomassa opened the pull request:
New version of the imaging pipeline that is compatible with the official L1 fits files (it is still compatible with the L1A fits files). The whole pipeline has beed reorganized so that appropriate structures are created. In summary, the following main changes have been applied:

1. Science and background L1(A) fits files are read by means of 'stx_read_pixel_data_fits_file.pro', so that imaging and spectroscopy share the same routine for reading data.
2. The coordinates ('mapcenter' and 'xy_flare') provided as input to the imaging pipeline are now conceived in the STIX reference frame. Every coordinate transformation has to be done before entering the pipeline. Procedures for coordinate transformations have been implemented (see procedures in '/stix/idl/processing/coordinates').
3. Selection of the time and energy intervals is slightly different compared to the previous version. In the updated version, we select the smallest time and energy intervals that contain the ones provided as input. The lower and upper edges of the selected time and energy intervals (potentially different from those provided as input) are saved in the pipeline structures (see point 7). Due to these modifications, if we use the same fits files and the same time and energy intervals with the previous and the current version, we may obtain different results.
4. A new function for computing the transmission function of a subcollimator have been implemented ('stx_subc_transmission'). This will be adapted to be compliant with the latest computations by Gordon.
5. Images can now be reconstructed also using all pixels (sumcase="ALL") or small pixels only (sumcase="SMALL"). 
6. Routines for plotting light curve, Moire patterns and visibility amplitude vs resolution have been adapted.
7. The pipeline works as follows: 
 - A science L1(A) fits file (and optionally also a background fits file) is read and a 'stx_pixel_data_structure' is created (containing the 32x12 matrix with raw counts);
- By using 'stx_sum_pixel_data', we sum the counts measured by the pixels. The resulting matrix is 32x4 and is saved in a 'stx_pixel_data_summed' structure;
- 'stx_pixel_data_summed2visibility' reads a 'stx_pixel_data_summed' structure and computes the associated (uncalibrated) visibility values, which are saved in a 'stx_visibility' structure;
-  An uncalibrated 'stx_visibility' structure is calibrated (i.e. the corresponding amplitude and phases are calibrated) with 'stx_calibrate_visibility'.

The following changes need to be performed before merging:

1. Check 'stx_make_map': this function should be used at the end of each imaging method for creating a map structure (so far it is used only at the end of "MEM_GE" since we have to discuss about it; once it will be ready, we'll use it for all methods). The visibility structure that has been used for reconstructing the map is save in the map structure, so that visibility amplitudes and phase fitting plots can be displayed (by means of 'stx_plot_fit_map').
2. Check 'stx_vis_fwdfit_pso':  the parameter values that are printed at the end of the optimization step are potentially conceived in a "wrong" reference system. @annavolp, could you please check it?
3. 'stx_create_auxiliary_data' has to be updated as discussed with @afbattaglia and @FredSchuller (5 min of integration, unless specified by the user).
4. Remove procedures of the previous version (e.g. 'stix2vis_sep2021'). I left them for now just for testing purposes
5. Test 'stx_subc_transmission' more in detail.
6. Test the use of small pixels only for image reconstruction more in detail.

paolomassa commented:
@grazwegian, please check the updated version of 'stx_read_pixel_data_fits_file'. This should fix the issue with the pixel mask of the L1 fits file, which has a different dimension w.r.t. that of the L1A fits files

paolomassa commented:
Please, also note that the imaging demo has a different name (now it is 'stx_imaging_demo') and that, similarly to the spectroscopy demo, there is a folder ('stix/dbase/demo/imaging') which contains the data once downloaded by the demo.

FredSchuller commented:
Hi @paolomassa 
Two questions about this new version:

- you write "2. The coordinates ('mapcenter' and 'xy_flare') provided as input to the imaging pipeline are now conceived in the STIX reference frame". Do I understand right that we are back to the previous version, where the user has to input coordinates in the STIX (rotated) frame, and the software makes the transformation to solar coordinates?

- do you plan to include your function "stx_estimate_location" at some point?

paolomassa commented:
Hi @FredSchuller, answers below:

1. In this new version of the software, the imaging pipeline takes as input coordinates conceived in the STIX (rotated) reference frame. Hence, the user have to preform all the transformations outside the pipeline. Operationally, the user can determine the 'mapcenter' and 'xy_flare' coordinates in the SOLO HPC coordinate frame. Then, by means of 'stx_hpc2stx_coord', the user can transform those coordinates into the corresponding ones in the STIX (rotated) reference frame, and provide the transformed coordinates as input to the imaging pipeline. Please, refer to the updated 'stx_imaging_demo' for details
2. Yes, I am planning to include 'stx_estimate_location', not sure to be able to do this in the next release

afbattaglia commented:
Hi all,
Paolo, brilliant work, really!! I tested the software in two ways:  
- firstly, I tested the demo (stx_imaging_demo) with this new version;
- secondly, I adapted some of my scripts to this new version and compared the results with the ones obtained with the old software.
Here attached you can find some comments regarding the two steps. It works very nicely, but there are still a few things that we should maybe discuss or double check (I do not exclude anyway that I wrongly used the software), e.g., the correction for the short time binning and the energy shift due to calibration issues (see pdf for more details).

[afbattaglia_Test-Paolos-new-imaging-software.pdf](https://github.com/i4Ds/STIX-GSW/files/9606597/afbattaglia_Test-Paolos-new-imaging-software.pdf)

paolomassa commented:
Hi @afbattaglia ,

thanks for the feedback on the code! I quickly reply below to some of your comments, I'll go through them more carefully in the next days:

1. As for the 'setenv' issue, I think we should have the same problem in the spectroscopy demo (see e.g., line 35 of 'stx_ospex_spectroscopy_demo.pro'). Maybe we should discuss with @grazwegian and address this problem in the same way for both the spectroscopy and imaging routines.
2. Thanks for reporting the change of the name of the L2 fits files, I'll correct it in the demo.
3. As for the use of the official L1 fits files in the demo, you are right, we should probably download them. I'll correct it in the demo. A general question about this: for every L1A fits file, is there a corresponding L1 fits file? In that case, is the UID the same?
4. As for the energy shift due to calibration issues, this shift is applied by 'stx_read_pixel_data_fits_file.pro', which has been originally created for spectroscopy. @grazwegian, do you know whether the shift is reasonable or not for the May 22, 2021 analyzed by @afbattaglia ?
5. As for short time binning issue, try to set the keyword 'shift_duration', it should work. If you set it in the routine that creates the visibilities, it should propagate through the '_extra' keyword and it should be used in 'stx_read_pixel_data_fits_file.pro'. If needed I will add a specific keyword for this.
6. I'm working to include the 'stx_estimate_location' routine in this release, it should be ready soon

Thanks again!

afbattaglia commented:
Thanks Paolo for addressing all these points!

Regarding point 5 of your list (short time binning), I checked the _shift_duration_ keyword and it works. The only thing is that the _shift_duration_ within the _extra propagates also when using **stx_read_pixel_data_fits_file** for reading the background file (see line 98 of **stx_construct_pixel_data**). Since in this case it is only a single time step, it fails. A solution would be to explicitly have _shift_duration_ in the set of keywords and not in the _extra and then avoid passing it for the background. What do you think? Do you have a better solution?

Regarding point 4 (energy shift), yes, we would need @grazwegian input there.

From my side, once these two issues are fixed, I can approve the request. Again, excellent work Paolo!

FredSchuller commented:
Hi @paolomassa , @afbattaglia , all,
I also ran the latest version of the imaging demo and it mostly worked fine. The only issue was with
`getenv('stx_det')`
in _stx_date2elut_file.pro_, because this is case-sensitive and I had defined STX_DET in my shell.

There are still several issues that need to be fixed before making a new official release. But it would be good to merge this pull request first, so that I can make changes and open a new pull request.

The following changes can then be done quickly:
 - in stx_create_auxiliary_data, the result is wrong if there is no measurement within the input time range in the AUX file (e.g. when using a very short time range, less than 64s, which falls between two measurements). I already implemented an interpolation in my local version, I can include that in the next pull request
 - it would really be useful to include the function _stx_estimate_location_
 - it would also be good to display some informative messages, e.g. what is the pointing used (and where it comes from)

Regarding the last point, this is a somewhat bigger issue on my side: we need to find a way to make use of the aspect solution if and only if it is reliable (this is also described in issue #110 . That will require more work but this is important before the next release.

afbattaglia commented:
Hi @paolomassa,
I just noticed the following. I was trying to reconstruct an image with one energy bin only, from 5 to 6 keV, but the map tells me that it used 4 to 6 keV. I think we had a similar discussion also with the old software, hadn't we? Is it not possible then to reconstruct images with one energy bin only?

paolomassa commented:
Hi @FredSchuller, @afbattaglia,

- As for the issues with 'getenv' you both reported: I checked, and in 'stix/setup.stix_env' the environmental variables are defined both in uppercase and lowercase to avoid problems with case-sensitive operative systems (in theory). Should we use just the uppercase then?
- I'm working on the 'stx_estimate_location' procedure; it should be ready soon.
- @FredSchuller, could you please modify 'stx_create_auxiliary_data' as you discussed today at the POST NECP discussion meeting? Maybe we could have a separate meeting/email exchange on this. I'll write you soon
- @afbattaglia, could you please give me more details on the test you performed (e.g., event considered, imaging algorithm used, etc.)? Below an example in which I reconstruct the June 7, 2020. event with MEM_GE:

https://drive.google.com/file/d/1UOc0Gzl4lpCwGjVe-gKJ3dV4IwFt8LjE/view?usp=sharing

 I integrate in the energy range 5 - 6 keV and the energy range saved in the map is still 5 - 6 keV. Sorry, I can't understand where the problem is.

paolomassa commented:
@afbattaglia I also added the keyword 'shift_duration' in such a way that it is not passed when we read the background file. Please, try it, it should work now. Thanks for your feedback!

afbattaglia commented:
Hi @paolomassa ,

Thanks for testing. Yes, indeed I looked at other flares and this seems to be fine. However, if I give as input energy_range = [6,7] for the SOL20210829T10:10 flare (file ID: 2108290133) it seems that it takes 5-7 keV. Is this maybe due to the -1.6 keV shift? 

Regarding the /shift_duration keyword, yes, it works now! Thanks a lot

paolomassa commented:
Hi @afbattaglia, 

if you have a shift in energy of -1.6, that's the reason behind the selection of the energy interval ~5-7 keV. Indeed, due to the shift, the edges of the energy bins become 2.4, 3.4, 4.4, 5.4, 6.4, 7.4 and so on. The new version of the software selects the smallest energy range that contains the one that you provide as input (6-7 keV); hence, in this case, it selects from 5.4 to 7.4 keV. Then the lower and upper edge of the energy interval that is actually selected by the code are saved in the visibility structure and passed to the final map

afbattaglia commented:
Great!! Thanks @paolomassa !

FredSchuller commented:
Hi @paolomassa 

> * @FredSchuller, could you please modify 'stx_create_auxiliary_data' as you discussed today at the POST NECP discussion meeting? Maybe we could have a separate meeting/email exchange on this. I'll write you soon

I tried to do so... I have cloned your branch _imaging_pipeline_ locally, then I made quite some changes in _stx_create_auxiliary_data_ (and a few minor changes in other files, but nothing dramatic). I could commit my changes but _git push_ did not work:
`remote: Permission to paolomassa/STIX-GSW.git denied to FredSchuller.
fatal: unable to access 'https://github.com/paolomassa/STIX-GSW.git/': The requested URL returned error: 403
`

Not sure if that's because the pull request has to be merged first? or somehow you're the only one who can make changes?

paolomassa commented:
Hi @FredSchuller 

I invited you as a collaborator to my repository. I hope now you have the rights to make a push. Let me know

FredSchuller commented:
Thanks @paolomassa ! It looks like it worked now...

grazwegian commented:
> Hi @afbattaglia ,
> 
> thanks for the feedback on the code! I quickly reply below to some of your comments, I'll go through them more carefully in the next days:
> 
> 1. As for the 'setenv' issue, I think we should have the same problem in the spectroscopy demo (see e.g., line 35 of 'stx_ospex_spectroscopy_demo.pro'). Maybe we should discuss with @grazwegian and address this problem in the same way for both the spectroscopy and imaging routines.
> 2. Thanks for reporting the change of the name of the L2 fits files, I'll correct it in the demo.
> 3. As for the use of the official L1 fits files in the demo, you are right, we should probably download them. I'll correct it in the demo. A general question about this: for every L1A fits file, is there a corresponding L1 fits file? In that case, is the UID the same?
> 4. As for the energy shift due to calibration issues, this shift is applied by 'stx_read_pixel_data_fits_file.pro', which has been originally created for spectroscopy. @grazwegian, do you know whether the shift is reasonable or not for the May 22, 2021 analyzed by @afbattaglia ?
> 5. As for short time binning issue, try to set the keyword 'shift_duration', it should work. If you set it in the routine that creates the visibilities, it should propagate through the '_extra' keyword and it should be used in 'stx_read_pixel_data_fits_file.pro'. If needed I will add a specific keyword for this.
> 6. I'm working to include the 'stx_estimate_location' routine in this release, it should be ready soon
> 
> Thanks again!



> Hi @afbattaglia ,
> 
> thanks for the feedback on the code! I quickly reply below to some of your comments, I'll go through them more carefully in the next days:
> 
> 1. As for the 'setenv' issue, I think we should have the same problem in the spectroscopy demo (see e.g., line 35 of 'stx_ospex_spectroscopy_demo.pro'). Maybe we should discuss with @grazwegian and address this problem in the same way for both the spectroscopy and imaging routines.
> 2. Thanks for reporting the change of the name of the L2 fits files, I'll correct it in the demo.
> 3. As for the use of the official L1 fits files in the demo, you are right, we should probably download them. I'll correct it in the demo. A general question about this: for every L1A fits file, is there a corresponding L1 fits file? In that case, is the UID the same?
> 4. As for the energy shift due to calibration issues, this shift is applied by 'stx_read_pixel_data_fits_file.pro', which has been originally created for spectroscopy. @grazwegian, do you know whether the shift is reasonable or not for the May 22, 2021 analyzed by @afbattaglia ?
> 5. As for short time binning issue, try to set the keyword 'shift_duration', it should work. If you set it in the routine that creates the visibilities, it should propagate through the '_extra' keyword and it should be used in 'stx_read_pixel_data_fits_file.pro'. If needed I will add a specific keyword for this.
> 6. I'm working to include the 'stx_estimate_location' routine in this release, it should be ready soon
> 
> Thanks again!

1. I am puzzled why this is not working for case sensitive systems as the the setup file includes both. It looks like other instruments just use uppercase in their files so maybe we should switch to the same. 

4. There was initially a typo in stx_check_energy_shift where the time range where the automatic shift was applied was incorrect. This was resolved last month so it would be good to check the version you are running is up to date. The shift was only seen during the late 2020 checkout so should not be applied to a flare on May 22, 2021.

grazwegian commented:
@paolomassa does stx_subc_transmission replace stix_gtrans32_test for calculating the transmission through the grids? If so I should update on the spectroscopy side.

paolomassa commented:
Hi @grazwegian
yes, 'stx_subc_transmission' would replace 'stix_gtrans32_test'. For now, it is conceived only for single layer grids (i.e., for subcollimators 3 to 10). I tested it and it seems to work; however, please, double-check it

paolomassa commented:
Hi @grazwegian, @samaloney 

do you have any comments about the routine 'stx_make_map'? Can I use at the end of every imaging method for creating a stix map which has the same information for every algorithm? My main concern is that, in the map structure, I save both the visibility values used for reconstructing the map and those predicted from the map before rotations (when the map is in the STIX reference frame). I need that for computing the data fitting plots; however, I do not know whether this is compliant with general policy about maps.
Please, let me know what do you think about this so that I will make the appropriate modifications by the end of this week. Thanks!

grazwegian commented:
> Hi @grazwegian
> yes, 'stx_subc_transmission' would replace 'stix_gtrans32_test'. For now, it is conceived only for single layer grids (i.e., for subcollimators 3 to 10). I tested it and it seems to work; however, please, double-check it

Hi. I had a look at stx_subc_transmission but for one of the finer grids (Number: 13, Index:12, Label: 1b ) I get a value for the transmission though for the others it is set to 1 - I am guessing it should also give a value of 1 for the moment? 

nontop24 = [9,10,11,12,13,17,18,19]
grid_factors=stx_subc_transmission([0.,0.])
grid_factors[nontop24-1]
       1.0000000       1.0000000       1.0000000       1.0000000      0.55560583       1.0000000       1.0000000       1.0000000

paolomassa commented:
Hi @grazwegian 

yes you are right. I don't know what's happening in detector 1b, let me check this. Thanks for spotting this!

paolomassa commented:
Hi @grazwegian 

I corrected a small bug and I fixed the issue you raised about the transmission of detector 1b

paolomassa commented:
Thanks! Also, in the current version of the imaging pipeline, I kept the old functions for creating a visibility structure (e.g., 'stix2vis_sep2021') for testing purposes. Should I keep them for facilitating the transition or it is better to remove everything that won't be used anymore?

paolomassa commented on stix/idl/processing/coordinates/stx_hpc2rtn_coord.pro:
```diff
@@ -0,0 +1,67 @@
+;
+; INPUTS:
+;
+;   xy_coord: bi-dimensional array of coordinates (arcsec). If the 'inverse' keyword is set to 0,
+;             the coordinates are assumed to be in the Helioprojective Cartesian coordinate frame
```
Yes, I will correct it

paolomassa commented on stix/idl/processing/pixel_data/stx_construct_pixel_data.pro:
```diff
@@ -0,0 +1,404 @@
+
+  endfor
+  
+  if keyword_set(path_bkg_file) then begin
+    
+    for i=0,31 do begin
+      
+      counts_bkg[*,i]       = counts_bkg[*,i]/subc_transmission_bkg[i]*0.25
+      counts_error_bkg[*,i] = counts_error_bkg[*,i]/subc_transmission_bkg[i]*0.25
+      
+    endfor
+    
+  endif
+
+endif
+
```
I agree with your comment, I will correct it

samaloney commented:
> Hi @grazwegian, @samaloney
> 
> do you have any comments about the routine 'stx_make_map'? Can I use at the end of every imaging method for creating a stix map which has the same information for every algorithm? My main concern is that, in the map structure, I save both the visibility values used for reconstructing the map and those predicted from the map before rotations (when the map is in the STIX reference frame). I need that for computing the data fitting plots; however, I do not know whether this is compliant with general policy about maps. Please, let me know what do you think about this so that I will make the appropriate modifications by the end of this week. Thanks!

I'm not sure I follow is the problem you have data from both before and after rotation in the map structure or that you have the visibility values on the map structure? If it the latter I don't think it's an issue extra data in the map structure should be a problem.

I do think we need to work on the keywords in the fits file the is created when saving the created maps, they can't be opened by displayed properly by standard fits map viewers e.g. sunpy, ds9

samaloney commented on stix/idl/processing/imaging/stx_estimate_flare_location.pro:
```diff
@@ -0,0 +1,121 @@
+
+;;******* Determine the pixel size from the apparent radius of the Sun for computing a full-disk Back Projection map
+rsun = aux_data.RSUN
+pixel = rsun * 2.6 / imsize
```
whey 2.6 maybe just add comment?

paolomassa replied:
Done, I'll push soon

samaloney replied:
Will this be fixed here or should we make a new issue?

paolomassa replied:
I added a comment in the code

samaloney commented on stix/idl/processing/imaging/stx_plot_fit_map.pro:
```diff
@@ -0,0 +1,176 @@
+;
+; PURPOSE:
+;
+;   Read a STIX science L1 fits file (and potentially a STIX background L1 fits file) and contruct a
+;   calibrated 'stx_visibility' structure
```
Seems different to what the function does?

paolomassa replied:
Yes, thank, I'll correct this

samaloney replied:
So will we open an issue for this or will it be fixed in this PR?

samaloney commented on stix/idl/processing/pixel_data/stx_construct_pixel_data.pro:
```diff
@@ -0,0 +1,393 @@
+if energy_range[0] gt energy_range[1] then message, "Energy range lower edge is greater than the higher edge"
+
+
+stx_read_pixel_data_fits_file, path_sci_file, data_str = data, t_axis = t_axis, e_axis = e_axis, alpha = alpha, shift_duration=shift_duration, _extra=extra
```
what `alpha` L1A v L1?

grazwegian replied:
Yes, the alpha keyword means the file in question is L1A rather than L1

samaloney commented on stix/idl/structures/stx_make_map.pro:
```diff
@@ -0,0 +1,123 @@
+;   paolo.massa@wku.edu
+;-
+
+function stx_make_map, im_map, aux_data, pixel, method, vis
```
What the frame of the input stx_rtn? maybe an option to  not rotate

paolomassa replied:
The input image is conceived in the STIX coordinate frame (see line 18). I would not add the option not to rotate for now, it needs to be discussed more

samaloney reviewed (commented):
Great work by all! I think it could be merged as is, but would appreciate some feed back on comments.

samaloney commented on stix/idl/processing/imaging/stx_estimate_flare_location.pro:
```diff
@@ -0,0 +1,121 @@
+
+
+;;******* Use Giordano's (u,v) points: no need to perform projection correction (see Giordano et al., 2015)
+subc_str = stx_construct_subcollimator()
```
Does this not give back normal u, v points - so not Giordano's?

paolomassa replied:
Lines 69-75 create a visibility structure and change the definition of the (u,v) points from the original ones to Giordano's ones. Giordano's (u,v)-points do not require projection correction and, hence, do not need an a priori estimate of the flare location, which is determined by this procedure.
In the future, we will use Giordano's (u,v)-points by default, but we need to discuss more on this

samaloney replied:
Sorry my bad I see now.

FredSchuller commented on stix/idl/processing/aux_data/stx_create_auxiliary_data.pro:
```diff
@@ -0,0 +1,197 @@
+;-
+function stx_create_auxiliary_data, fits_path, time_range, use_sas=use_sas, dont_use_sas=dont_use_sas, silent=silent
+
+  default, use_sas, 0
+  default, dont_use_sas, 0
```
Sure, I can rename these keywords as you suggest if it sounds clearer.
I just tested, if both are sets, then `dont_use_sas` wins. We could display a warning since it should not happen to set both.

FredSchuller commented on stix/idl/processing/imaging/fit_map_2dgauss.pro:
```diff
@@ -0,0 +1,53 @@
+  ; Fit a 2D-Gaussian to the data
+  res = gauss2dfit(in_map.data, coeff, /tilt)
+  ; Here are the center coordinates
+  gauss_X = in_map.xc + in_map.dx*(coeff[4] - nbX/2.)
+  gauss_Y = in_map.yc + in_map.dY*(coeff[5] - nbY/2.)
```
The only assumptions is that there is one source that dominates and that can be relatively well fitted by a 2D Gaussian, or do you mean anything else?
Anyway, it's not really part of the imaging pipeline, but since I'm using it a lot, I found it convenient to include it here. I can remove it and keep it in my "private" folder if you wish.

samaloney replied:
I was speaking about the conversion of the pixel to world coordinates it probably fine but `in_map.xc + in_map.dx(pixelx_coord)` isn't always valid

paolomassa commented:
> > Hi @grazwegian, @samaloney
> > do you have any comments about the routine 'stx_make_map'? Can I use at the end of every imaging method for creating a stix map which has the same information for every algorithm? My main concern is that, in the map structure, I save both the visibility values used for reconstructing the map and those predicted from the map before rotations (when the map is in the STIX reference frame). I need that for computing the data fitting plots; however, I do not know whether this is compliant with general policy about maps. Please, let me know what do you think about this so that I will make the appropriate modifications by the end of this week. Thanks!
> 
> I'm not sure I follow is the problem you have data from both before and after rotation in the map structure or that you have the visibility values on the map structure? If it the latter I don't think it's an issue extra data in the map structure should be a problem.
> 
> I do think we need to work on the keywords in the fits file the is created when saving the created maps, they can't be opened by displayed properly by standard fits map viewers e.g. sunpy, ds9 see #97

Hi @samaloney 

my concern was about the fact that I added more keywords to the map structure. However, as you replied, it should not be a problem

paolomassa commented:
Dear All,

some comments below:

- @FredSchuller, could you please comment '-1' factors in 'stx_create_auxiliary_data'? I do not remember the details behind these factors
- Do we still want to keep the old version of the imaging pipeline to facilitate the transition? If not, then I could remove a bunch of old procedures
- I like the idea of moving 'Mapcenter_correction_factors.csv' to a dedicated folder in dbase. However, if we do this, the  old version of the imaging pipeline should not work anymore (see point above)

samaloney commented on stix/idl/processing/aux_data/stx_create_auxiliary_data.pro:
```diff
@@ -0,0 +1,197 @@
+;************** Get the indices corresponding to the considered time range
+this_time_range = anytim(time_range)
+time_data       = anytim(aux_data_str.TIME_UTC)
+time_data      -= 32.   ; shift AUX data by half a time bin (starting time vs. bin centre)
```
Yea so the spice data is calculated at the time in the fits file no need for the -32 (which can change too). @FredSchuller are you shifting the time in IDL for the SAS pointing calculation

FredSchuller replied:
@samaloney
No I don't apply any shift on my side, but the problem is: the variations in the aspect signals do not perfectly match what I expect from the as-flown attitude (as obtained from the spice kernels). My interpretation is that the time stamps that are written in the L1 HK FITS files refer to the middle of the time bin, while the data samples (at least for aspect) were acquired at the beginning of the time bin.
Also when comparing aspect burst mode data with regular data covering the same time range, I see a shift consistent with 32s - I am now preparing a figure to illustrate that.

samaloney replied:
@nicHoch what do think about this?

FredSchuller replied:
![burst_vs_HK_20220613](https://user-images.githubusercontent.com/36623402/198065745-2af0d5da-0460-4a80-96e3-9d1e5de95a60.png)
Here's an example showing burst mode data (blue line) and regular HK data (red open squares)

samaloney replied:
See #115 I think the only change needed here is to no hard code 32s but use 1/2 the time between bins

samaloney replied:
See issue #121

samaloney commented:
> @paolomassa @samaloney
> Regarding the '-1' factors in 'stx_create_auxiliary_data' for yaw (but not for pitch), I guess this is required because of the way the auxiliary files are generated, e.g. if the spice kernels are used with a different reference frame compared to the one that we need to express the pointing in the STIX frame.
> But that's implemented somewhere in STIXcore, therefore I don't know the details.

Here is the relevant code we use the in particular line 303

https://github.com/i4Ds/STIXCore/blob/cc21dc25198a550fc162b1d3268403c404644a4c/stixcore/ephemeris/manager.py#L299-L321

FredSchuller commented:
> Here is the relevant code we use the in particular line 303

> https://github.com/i4Ds/STIXCore/blob/cc21dc25198a550fc162b1d3268403c404644a4c/stixcore/ephemeris/manager.py#L299-L321


OK, well... I could dig into the documentation of spiceypy to try and understand what is actually computed, but this should not delay the closing of this pull request (and the release of the GSW) any longer...

paolomassa commented:
Hi all,

I removed the old routines for computing the visibilities and I removed the dbase/demo/vis_demo folder (the files that we keep using have been moved to other folders). This is my last commit if no other modifications are required by you. There is still to discuss the coordinate transformations outside the routines for computing the visibilities:

mapcenter = stx_hpc2stx_coord(flare_loc, aux_data)
xy_flare  = mapcenter

It is probably confusing for non-expert users, however I will leave the code as it is for now so that we do not delay the release

samaloney commented:
I think the open discussions need to be resolved before we can merge this can be done by:
1. Hit the resolve button if the issue has been addressed either by code changes or agreement on the discussion
2. Creating a new issue if the discussion can't be resolved or needs further analysis

paolomassa commented on stix/idl/processing/aux_data/stx_create_auxiliary_data.pro:
```diff
@@ -0,0 +1,197 @@
+; If aspect solution is available (i.e., ~NaN) and is reliable (i.e. not much different from the spacecraft estimate) 
+; use it. Otherwise, use spacecraft pointing estimate
+
+readcol, loc_file( 'Mapcenter_correction_factors.csv', path = getenv('STX_VIS_DEMO') ), $
```
I made this change in the last cleanup commit

samaloney commented on stix/idl/processing/coordinates/stx_solo2stx_coord.pro:
```diff
@@ -0,0 +1,60 @@
+; PURPOSE:
+;
+;   Transform a two-element array of coordinates from the Solar Orbiter coordinate frame to the STIX coordinate frame 
+;   and viceversa. Essentially, it performs a 90 deg rotation as STIX is mounted on the side of the spacecraft
```
Created new issue #117

samaloney commented on stix/idl/processing/coordinates/stx_rtn2stx_coord.pro:
```diff
@@ -0,0 +1,65 @@
+;
+; NAME:
+;
+;   stx_rtn2stx_coord
```
Created new issue #118

samaloney commented on stix/idl/processing/coordinates/stx_rtn2solo_coord.pro:
```diff
@@ -0,0 +1,60 @@
+;
+; PURPOSE:
+;
+;   Transform a two-element array of coordinates from Solar Orbiter - Sun Radial-Tnagential-Normal coordinate frame
+;   to the Solar Orbiter coordinate frame and viceversa. Essentially, it performs a correction for the instrument pointing
```
Created new issue #119

samaloney commented on stix/idl/processing/aux_data/stx_create_auxiliary_data.pro:
```diff
@@ -0,0 +1,197 @@
+  YAW = ((time_data[t_after]-time_middle) * aux_data_str[t_before].ROLL_ANGLE_RPY[2] + $
+         (time_middle-time_data[t_before]) * aux_data_str[t_after].ROLL_ANGLE_RPY[2]) / $
+         (time_data[t_after]-time_data[t_before])
+  YAW = -1.*float(YAW) * 3600.
```
See #120

samaloney commented on stix/idl/processing/aux_data/stx_create_auxiliary_data.pro:
```diff
@@ -0,0 +1,197 @@
+  ;Roll angle (degrees)
+  ROLL_ANGLE = average(aux_data_str[time_ind].ROLL_ANGLE_RPY[0])
+  ;Yaw (arcsec)
+  YAW = -average(aux_data_str[time_ind].ROLL_ANGLE_RPY[2] * 3600.)
```
See #120

samaloney commented on stix/idl/processing/aux_data/stx_create_auxiliary_data.pro:
```diff
@@ -0,0 +1,197 @@
+
+  diff_ptg = norm(sas_pointing - spacecraft_pointing)
+
+  if diff_ptg lt 200. or use_sas then begin
```
See #122

samaloney reviewed (approved):
Moved all open discussions to new issues
